subprojects/logging/src/main/java/org/gradle/internal/logging/console/ProgressBar.java

```diff
@@ -60,6 +64,16 @@ public void moreProgress(int totalProgress) {
 
     public void update(boolean failing) {
         this.current++;
+        if (current > total) {
+            new Thread(new Runnable() {
+                @Override
+                public void run() {
+                    // do not do this directly or a deadlock happens
+                    // to prevent that deadlock, execute it separately in a new thread
+                    LOGGER.warn("More progress was logged than there should be ({} > {})", current, total);
```

Comment:
I'm trying to wrap my head around how deferring stops us having a livelock here.
IIUC `LOGGER` will still report the event, and it should make a loop.

Reply:
If you merge this PR with #24592, comment out the deadlock prevention code, and use my reproducer from #24370, you can easily reproduce the deadlock.
Here the thread dump of the two relevant threads that block each other, I just reproduced it:
```threaddump
"main@1" prio=5 tid=0x1 nid=NA waiting for monitor entry
  java.lang.Thread.State: BLOCKED
	 blocks pool-1-thread-1@1771
	 waiting for pool-1-thread-1@1771 to release lock on <0xa8c> (a java.lang.Object)
	  at org.gradle.internal.logging.console.ThrottlingOutputEventListener.onOutput(ThrottlingOutputEventListener.java:79)
	  at jdk.internal.reflect.GeneratedMethodAccessor1.invoke(Unknown Source:-1)
	  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	  at java.lang.reflect.Method.invoke(Method.java:566)
	  at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	  at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	  at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:43)
	  at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:257)
	  at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:164)
	  at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:148)
	  at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	  at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	  at com.sun.proxy.$Proxy0.onOutput(Unknown Source:-1)
	  at org.gradle.internal.logging.sink.OutputEventTransformer.invokeListener(OutputEventTransformer.java:113)
	  - locked <0xa80> (a java.lang.Object)
	  at org.gradle.internal.logging.sink.OutputEventTransformer.onOutput(OutputEventTransformer.java:90)
	  at org.gradle.internal.logging.sink.OutputEventRenderer.onOutput(OutputEventRenderer.java:430)
	  at org.gradle.internal.logging.sink.OutputEventListenerManager$1.onOutput(OutputEventListenerManager.java:36)
	  at org.gradle.launcher.daemon.client.DaemonClient.monitorBuild(DaemonClient.java:242)
	  at org.gradle.launcher.daemon.client.DaemonClient.executeBuild(DaemonClient.java:202)
	  at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:152)
	  at org.gradle.launcher.daemon.client.DaemonClient.execute(DaemonClient.java:98)
	  at org.gradle.launcher.cli.RunBuildAction.run(RunBuildAction.java:57)
	  at org.gradle.internal.Actions$RunnableActionAdapter.execute(Actions.java:167)
	  at org.gradle.launcher.cli.DefaultCommandLineActionFactory$ParseAndBuildAction.execute(DefaultCommandLineActionFactory.java:259)
	  at org.gradle.launcher.cli.DefaultCommandLineActionFactory$ParseAndBuildAction.execute(DefaultCommandLineActionFactory.java:230)
	  at org.gradle.launcher.cli.DebugLoggerWarningAction.execute(DebugLoggerWarningAction.java:74)
	  at org.gradle.launcher.cli.DebugLoggerWarningAction.execute(DebugLoggerWarningAction.java:30)
	  at org.gradle.launcher.cli.WelcomeMessageAction.execute(WelcomeMessageAction.java:96)
	  at org.gradle.launcher.cli.WelcomeMessageAction.execute(WelcomeMessageAction.java:40)
	  at org.gradle.launcher.cli.NativeServicesInitializingAction.execute(NativeServicesInitializingAction.java:44)
	  at org.gradle.launcher.cli.NativeServicesInitializingAction.execute(NativeServicesInitializingAction.java:26)
	  at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:41)
	  at org.gradle.launcher.cli.ExceptionReportingAction.execute(ExceptionReportingAction.java:26)
	  at org.gradle.launcher.cli.DefaultCommandLineActionFactory$WithLogging.execute(DefaultCommandLineActionFactory.java:361)
	  at org.gradle.launcher.Main.doAction(Main.java:35)
	  at org.gradle.launcher.bootstrap.EntryPoint.run(EntryPoint.java:50)
	  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-1)
	  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	  at java.lang.reflect.Method.invoke(Method.java:566)
	  at org.gradle.launcher.bootstrap.ProcessBootstrap.runNoExit(ProcessBootstrap.java:60)
	  at org.gradle.launcher.bootstrap.ProcessBootstrap.run(ProcessBootstrap.java:37)
	  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-1)
	  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	  at java.lang.reflect.Method.invoke(Method.java:566)
	  at org.gradle.launcher.GradleMain.main(GradleMain.java:34)

"pool-1-thread-1@1771" prio=5 tid=0x12 nid=NA waiting for monitor entry
  java.lang.Thread.State: BLOCKED
	 blocks main@1
	 waiting for main@1 to release lock on <0xa80> (a java.lang.Object)
	  at org.gradle.internal.logging.sink.OutputEventTransformer.invokeListener(OutputEventTransformer.java:113)
	  at org.gradle.internal.logging.sink.OutputEventTransformer.onOutput(OutputEventTransformer.java:106)
	  at org.gradle.internal.logging.sink.OutputEventRenderer.onOutput(OutputEventRenderer.java:430)
	  at org.gradle.internal.logging.sink.OutputEventListenerManager$1.onOutput(OutputEventListenerManager.java:36)
	  at org.gradle.internal.logging.slf4j.OutputEventListenerBackedLogger.log(OutputEventListenerBackedLogger.java:52)
	  at org.gradle.internal.logging.slf4j.BuildOperationAwareLogger.log(BuildOperationAwareLogger.java:140)
	  at org.gradle.internal.logging.slf4j.BuildOperationAwareLogger.log(BuildOperationAwareLogger.java:154)
	  at org.gradle.internal.logging.slf4j.BuildOperationAwareLogger.log(BuildOperationAwareLogger.java:148)
	  at org.gradle.internal.logging.slf4j.BuildOperationAwareLogger.warn(BuildOperationAwareLogger.java:382)
	  at org.gradle.internal.logging.slf4j.OutputEventListenerBackedLogger.warn(OutputEventListenerBackedLogger.java:25)
	  at org.gradle.internal.logging.console.ProgressBar.update(ProgressBar.java:79)
	  at org.gradle.internal.logging.console.BuildStatusRenderer.phaseProgressed(BuildStatusRenderer.java:140)
	  at org.gradle.internal.logging.console.BuildStatusRenderer.onOutput(BuildStatusRenderer.java:106)
	  at org.gradle.internal.logging.console.AbstractUserInputRenderer.onOutput(AbstractUserInputRenderer.java:56)
	  at org.gradle.internal.logging.console.ThrottlingOutputEventListener.renderNow(ThrottlingOutputEventListener.java:105)
	  at org.gradle.internal.logging.console.ThrottlingOutputEventListener.onOutput(ThrottlingOutputEventListener.java:82)
	  - locked <0xa8c> (a java.lang.Object)
	  at org.gradle.internal.logging.console.ThrottlingOutputEventListener$1.run(ThrottlingOutputEventListener.java:66)
	  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	  at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	  at java.lang.Thread.run(Thread.java:829)
```
By not logging directly from the progress bar, but doing it from a different thread, this mutual deadlock is broken.

Reply:
Of course this deadlock could also be another bug that should instead be fixed properly so that it cannot happen, but I didn't investigate in that direction. :-)